compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2477,7 +2478,13 @@ impl<'a> Parser<'a> {
         let mut cond =
             self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL | Restrictions::ALLOW_LET, None)?;
 
-        CondChecker { parser: self, forbid_let_reason: None }.visit_expr(&mut cond);
+        let mut checker = CondChecker::new(self);
+        checker.visit_expr(&mut cond);
+        if checker.seen_missing_let {
```

Comment:
Instead of maintaining global state here, why don't we just bubble up a parse error to stop compilation, or taint the AST with an `ExprKind::Error`?